File: Templates/Union Find.cpp

// Implementation of Disjoint Set using Union-Find data structure optimized with both “path compression” and “union by rank”.
// Time Complexity 
// *  Constructor: O(N) 
// *  Find(): O(a(N)) ~ O(1)
// *  Union(): O(a(N)) ~ O(1)
// *  Connectivity:  O(a(N)) ~ O(1)
// Space Complexity: O(N)

class UnionFind {
private:
  vector<int> root;
  vector<int> rank;
public:
    UnionFind(int sz) : root(sz), rank(sz) {
        for (int i = 0; i < sz; i++) {
            root[i] = i;
            rank[i] = 1;
        }
    }
    int find(int x) {
        if (x == root[x]) {
            return x;
        }
        return root[x] = find(root[x]);
    }
    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            if (rank[rootX] > rank[rootY]) {
                root[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                root[rootX] = rootY;
            } else {
                root[rootY] = rootX;
                rank[rootX] += 1;
            }
        }
    }
    bool connected(int x, int y) {
        return find(x) == find(y);
    }
};

// reference: Leetcode

File: Templates/Python/Union_Find.py
class UnionFind:
    def __init__(self,size):
        self.root = [-1]*size
        self.rank = [1]*size
        
    def find(self,x):
        if self.root[x] == -1:
            return x
        self.root[x] = self.find(self.root[x])
        return self.root[x] 
    
    def union(self,x,y):
        rootX = self.find(x);
        rootY = self.find(y);
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX;
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY;
            else:
                self.root[rootY] = rootX;
                self.rank[rootX] += 1;
                
    def IsConnected(self,x, y):
        return self.find(x) == self.find(y)
            
